cuda/test/solver/gmres_kernels.cpp

```diff
@@ -0,0 +1,285 @@
+        ref, residual.get(), residual_norm.get(),
+        residual_norm_collection.get(), krylov_bases.get(),
+        final_iter_nums.get(), gko::solver::default_krylov_dim);
+    gko::kernels::cuda::gmres::initialize_2(
+        cuda, d_residual.get(), d_residual_norm.get(),
+        d_residual_norm_collection.get(), d_krylov_bases.get(),
+        d_final_iter_nums.get(), gko::solver::default_krylov_dim);
+
+    ASSERT_MTX_NEAR(d_residual_norm, residual_norm, 1e-14);
+    ASSERT_MTX_NEAR(d_residual_norm_collection, residual_norm_collection,
+                    1e-14);
+    ASSERT_MTX_NEAR(d_krylov_bases, krylov_bases, 1e-14);
+    gko::Array<gko::size_type> host_iter(ref, *d_final_iter_nums);
+    for (gko::size_type i = 0; i < final_iter_nums->get_num_elems(); ++i) {
+        ASSERT_EQ(host_iter.get_const_data()[i],
+                  final_iter_nums->get_const_data()[i]);
+    }
+}
+
+
+TEST_F(Gmres, CudaGmresStep1IsEquivalentToRef)
+{
+    initialize_data();
+    int iter = 5;
+
+    gko::kernels::reference::gmres::step_1(
+        ref, next_krylov_basis.get(), givens_sin.get(), givens_cos.get(),
+        residual_norm.get(), residual_norm_collection.get(), krylov_bases.get(),
+        hessenberg_iter.get(), b_norm.get(), iter, final_iter_nums.get(),
+        stop_status.get());
+    gko::kernels::cuda::gmres::step_1(
+        cuda, d_next_krylov_basis.get(), d_givens_sin.get(), d_givens_cos.get(),
+        d_residual_norm.get(), d_residual_norm_collection.get(),
+        d_krylov_bases.get(), d_hessenberg_iter.get(), d_b_norm.get(), iter,
+        d_final_iter_nums.get(), d_stop_status.get());
+
+    ASSERT_MTX_NEAR(d_next_krylov_basis, next_krylov_basis, 1e-14);
+    ASSERT_MTX_NEAR(d_givens_sin, givens_sin, 1e-14);
+    ASSERT_MTX_NEAR(d_givens_cos, givens_cos, 1e-14);
+    ASSERT_MTX_NEAR(d_residual_norm, residual_norm, 1e-14);
+    ASSERT_MTX_NEAR(d_residual_norm_collection, residual_norm_collection,
+                    1e-14);
+    ASSERT_MTX_NEAR(d_hessenberg_iter, hessenberg_iter, 1e-14);
+    ASSERT_MTX_NEAR(d_krylov_bases, krylov_bases, 1e-14);
+    gko::Array<gko::size_type> host_iter(ref, *d_final_iter_nums);
+    for (gko::size_type i = 0; i < final_iter_nums->get_num_elems(); ++i) {
+        ASSERT_EQ(host_iter.get_const_data()[i],
+                  final_iter_nums->get_const_data()[i]);
+    }
+}
+
+
+TEST_F(Gmres, CudaGmresStep2IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::kernels::reference::gmres::step_2(
+        ref, residual_norm_collection.get(), krylov_bases.get(),
+        hessenberg.get(), y.get(), x.get(), final_iter_nums.get(),
+        preconditioner.get());
+    gko::kernels::cuda::gmres::step_2(
+        cuda, d_residual_norm_collection.get(), d_krylov_bases.get(),
+        d_hessenberg.get(), d_y.get(), d_x.get(), d_final_iter_nums.get(),
+        d_preconditioner.get());
+
+    ASSERT_MTX_NEAR(d_y, y, 1e-14);
+    ASSERT_MTX_NEAR(d_x, x, 1e-14);
+}
+
```

Comment:
should we have a test comparing the solution to a linear system?

Reply:
Sure, I can add one, do you want me to hand-write the linear system, or also generate it randomly?

I think we previously decided against doing it because we would have to test a lot to get full coverage (all sparse formats, different preconditioner, etc.), but I am completely fine adding one test to compare the result between the reference and CUDA solution.

Reply:
I think we need the current test sizes because with CUDA, we need to run multiple thread blocks at the same time to check if that would fail (and multiple rhs). I think the CUDA implementation should run fine, it is the reference version that takes a long time (understandably, it should run correct and not necessarily fast).

@hartwiganzt The question is, what matrix should I test? Should I write one by hand and run both, or generate it randomly and hope it converges for both equally?

Reply:
@thoasm I recommend you use a finite difference discretization like poisson solver example https://github.com/ginkgo-project/ginkgo/blob/develop/examples/poisson_solver/poisson_solver.cpp

Reply:
Just to archive our decision here:
Because we do not have this test for any other solver, we omit it to stay consistent.
Additionally, we found the (biggest) problem with the unit tests after they had appropriate coverage.
For bigger which tests combination of multiple functions, we probably need [integration tests](https://en.wikipedia.org/wiki/Integration_testing) or [validation tests](https://en.wikipedia.org/wiki/Software_verification_and_validation).

We probably need to do validation tests before the release either way by hand to ensure everything works properly.